### PR TITLE
JW-660 vrouter_hyperv.py

### PR DESCRIPTION
**Pull request by m-kostrzewa**

### CODE REVIEW

**Comment by m-kostrzewa** [comment]
```
$guid = [guid]::NewGuid().guid
python .\vrouter_hyperv.py create --vm_location C:\snatvm --vhd_path C:\ubuntu-snat-vm.vhdx --mgmt_vswitch_name nat --vrouter_vswitch_name nat $guid $guid $guid
python .\vrouter_hyperv.py destroy $guid $guid $guid
```

**Comment by m-kostrzewa** [comment]
pip install netaddr requests

**Comment by Attil** [comment]
pip install paramiko

**Comment by Attil** [comment]
```PS C:\Users\Administrator\agent\controller\src\vnsw\opencontrail-vrouter-netns\opencontrail_vrouter_netns> python .\vrouter_hyperv.py create --vm_location D:\UbuntuNAT --vhd_path D:\ubuntu-snat-vm.vhdx --mgmt_vswitch_name int --vrouter_vswitch_name int $guid $guid $guid```
```Get-VM : Hyper-V was unable to find a virtual machine with name "contrail-wingw-af39cd8d-0d08-4449-af08-1e4ff9705cd5".
At line:1 char:1
+ Get-VM -Name contrail-wingw-af39cd8d-0d08-4449-af08-1e4ff9705cd5
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (contrail-wingw-...08-1e4ff9705cd5:String) [Get-VM], VirtualizationException
    + FullyQualifiedErrorId : InvalidParameter,Microsoft.HyperV.PowerShell.Commands.GetVM
```

**Comment by m-kostrzewa** [comment]
That error (debug )message is expected (script checks if vm with such name already exists by looking for error code when calling Get-VM) @Attil

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1]: do we even need dns?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1]: probably better to return error code

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1]: do we even need dns here?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: does anyone know? @sodar @Attil @mmasztalerczuk 

in vrouter_netns.py there is 
```
nic_left['mac'] = netaddr.EUI(self.args.vmi_left_mac,
                                          dialect=netaddr.mac_unix)```

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I'm not sure if the name here must be unique (if we require uuids here).

On Linux, these are `int-uuid1` and `gw-uuid2`. Here it's just `eth1` and `eth2`.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: meh :/ can we query hyper-v somehow? or maybe at the very least do ping polling?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Maybe we should ask Raja about this.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: This is duplicate code from vrouter_netns.py but we can't import that module on Windows. Should we consider refactoring both vrouter_netns and this?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: he said it should be NameSpacePort

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mclapinski) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Yes, we should.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mclapinski) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: remove space

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mclapinski) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: comment this out or something

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: You can remove the comment then.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: This is not used.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I suggest using `e.returncode` instead of parsing the string. Parsing the string isn't stable, while return codes should be consistent.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: This variable is misnamed, isn't it? The end result of this will not be (just) adapter name.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: It is certainly possible, as the VM will send some broadcast packets when the network interface is up. Is it feasible to do it though?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: It would be nice if both `AA-BB` and `aa:bb` were supported. One is used everywhere in Windows and another is used everywhere in Contrail.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1
Comment (Attil) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv_inject_ip.ps1]: No.

**Review by Attil — CHANGES_REQUESTED** [review]

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mclapinski) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: `first_available_ip = next(iter(free_ips))` should be the fastest way.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mclapinski) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I think you should decide whether you put `"""` in the same line or in a separate like. Like that for example:
```
"""
Create or destroy a Hyper-V Gateway VM for NAT between two virtual 
networks.
"""
```

**Review by mclapinski — CHANGES_REQUESTED** [review]

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (sodar) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: When Hyper-V internal switch is created, `ManagementOS` adapter with the same name as the switch is attached to this it as well. So I think we should use management switch name as a name for adapter.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (sodar) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Using `mac_eui48` as dialect will produce MAC address in following format: `DE-AD-BE-EF-00-00`, which is the standard Windows format. So probably when ports are registered in Contrail we should convert it to format used by Linux, i.e. with colons.

**Review by sodar — CHANGES_REQUESTED** [review]

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: This converts any MAC do the one needed by Windows (eui48)

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Hm... we'll need to figure this out when we start integrating it into agent itself

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/common.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/common.py]: A line should end with `\n`

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: ??

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Please, check my previous comment on this line.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Please, check my previous comment on this line.

**Review by Attil — CHANGES_REQUESTED** [review]

**Review by mclapinski — APPROVED** [review]

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (Attil) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: wouldn't retry_num < self.NUM_INJECT_RETRIES be more readable?

**Review by Attil — CHANGES_REQUESTED** [review]

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I think that you should add some checks here (problems with none). Maybe any unit tests?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I don't like this function. It is just kind of wrapper and it is *only* adding "powershell.exe" to cmds. You have to still add a lot of arguments when you want to invoke some command. The only adantage of this funtion is that you don't have to add three arguments to every invocation of subprocess. Also, there is no any error checking. 

This kind of design can create a lot of places where you can make mistake, or/and debugging this code is very hard. The solution maybe is more 'pythonic' but I think that it can create a lot of problems in future. 

I think that you should add the class which will be middleware between subprocess and your code. For every command which you want to execute from powershell. If you do something like that, then you can add more checks to verify your command (and as we now, we don't create simple few arguments command, but our commands are more complicated). I think that verifing how exactly look every command (and logging this to file) is very important for improving time of solving debugging problem.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I don't understand this **try**

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I think that you should divide this file into few files (modules)

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Maybe you should add this as config or arguments in main?

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py:82]: Stupid question - but why are you using single underscore? I think that you should use two or none.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I don't like init with some many arguments. I think that you should design this in different way (for example use dict as argument) https://stackoverflow.com/questions/5899185/class-with-too-many-parameters-better-design-strategy

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: If you create middleware for subprocess, then you could invoke this like powershell.addVMNetworkAdapter(VMName=self.wingw_name)

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (mmasztalerczuk) [inline src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I think that you should invoke this function when something went wrong also (because in other scenario we have to manually log into machine and delete/stop vm)

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py:82]: Trying to keep the convention from vrouter_netns.py (which is in the same directory)

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: As per our discussion, I renamed the function to `call_subprocess`

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I want to avoid doing this until we merge with 4

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: Refactored by moving some fields to other classes.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I agree that such class would be pretty useful, however I don't think it's worth the effort just for one script.

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I'm not sure about this, since this is not configurable in vrouter_netns.py either

File: src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py
Comment (m-kostrzewa) [reply src/vnsw/opencontrail-vrouter-netns/opencontrail_vrouter_netns/vrouter_hyperv.py]: I don't think unit tests for glue code are worth the effort tbh... if anything, I would do some integration tests against Hyper-V, but they aren't worth it either I think.

**Review by mmasztalerczuk — APPROVED** [review]

**Pull request merged**
